server/rest_slicer_cli.py

```diff
@@ -826,14 +829,163 @@ def genRESTEndPointsForSlicerCLIsInDocker(info, restResource, dockerImages):
+                        'string or a list of docker image strings')
+
+    if isinstance(dockerImages, list):
+        for img in dockerImages:
+            if not isinstance(img, str):
+                raise Exception('dockerImages must either be a single '
+                                'docker image string or a list of docker '
+                                'image strings')
+    else:
+        dockerImages = [dockerImages]
+
+    # create REST resource if given a name
+    if isinstance(restResource, str):
+        restResource = type(restResource,
+                            (Resource, ),
+                            {'resourceName': restResource})()
+
+    restResourceName = type(restResource).__name__
+
+    # Add REST routes for slicer CLIs in each docker image
+    cliList = []
+
+    for dimg in dockerImages:
+        # check if the docker image exists
+
+        id = localDockerImageExists(dimg)
+        if id is None:
+            id = pullDockerImage(dimg)
+        if id is None:
+            raise ValueError("The image %s does not exist locally or "
+                             "on the default registry" % dimg)
+        # get CLI list
+        cliListSpec = localDockerImageCLIList(dimg)
+        if cliListSpec is None:
+            raise ValueError("Could not get the cli list for the img %s most "
+                             "likely the docker image entrypoint does "
+                             "not accept the argument --list_cli" % dimg)
+        # pprint.pprint(cliListSpec)
+
+        cliListSpec = json.loads(cliListSpec)
 
+        # Add REST end-point for each CLI
+        for cliRelPath in cliListSpec.keys():
+            cliXML = localDockerImageclixml(dimg, cliRelPath)
+            if cliXML is None:
+                raise ValueError("Could not get the cli for the img %s cli %s"
+                                 " most likely the docker image entrypoint does"
+                                 " not accept the argument %s --xml" %
+                                 (dimg, cliRelPath, cliRelPath))
+            # create a POST REST route that runs the CLI
+            try:
+
+                # print cliXML
+                cliRunHandler = genHandlerToRunDockerCLI(dimg,
+                                                         cliRelPath,
+                                                         cliXML,
+                                                         restResource)
+            except Exception as e:
+                print "Failed to create REST endpoints for %s: %s" % (
+                    cliRelPath, e)
+                print "cli handler"
+                continue
+
+            cliSuffix = os.path.normpath(cliRelPath).replace(os.sep, '_')
+
+            cliRunHandlerName = 'run_' + cliSuffix
+            setattr(restResource, cliRunHandlerName, cliRunHandler)
+            restResource.route('POST',
+                               (cliRelPath, 'run'),
```

Comment:
Just to keep in mind (no need to change now), to get around name conflicts across docker images, maybe we can prefix this route with a massaged version (maybe use hash if there is no generic way to massage) of the docker image string